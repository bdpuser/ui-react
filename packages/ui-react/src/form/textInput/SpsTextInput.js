import React, { Component } from "react";
import classnames from "classnames";
import PropTypes from "prop-types";

const PRISTINE = "pristine";
const DIRTY = "dirty";
const ERROR = "error";
const VALID = "valid";

export default class SpsTextInput extends Component {
  constructor(props) {
    super(props);
    const { inputValue} = this.props;
    this.state = {
      interactionState: PRISTINE,
      validity: null,
      inputValue: inputValue
    };
    this.inputRef = React.createRef();
  }

  blurHandler = event => {
    const { blurCallback } = this.props;
    if (typeof blurCallback === "function") {
      blurCallback(event);
      // should this return out before running internal validation?
    }
    this.validateInput(event);
  };

  changeHandler = event => {
    this.setState((prev, props) => {
      if (prev.interactionState === PRISTINE) {
        return { interactionState: DIRTY };
      }
    });
    const { onChange } = this.props;
    if (event.target.value !== "") {
      this.setState({
        inputValue: event.target.value
      });
      if (typeof onChange === "function") {
        onChange(event.target.value);
      }
    }
  };

  validateInput = event => {
    const { required, customValidator } = this.props;
    // should we allow validation on non-required?
    if (!required || !event || !event.currentTarget) {
      return;
    }
    const input = event.target;
    let valid;
    if (typeof customValidator === "function") {
      valid = customValidator(String(input.value));
    }
    if (!customValidator || customValidator !== "function") {
      valid = input.value !== "" && input.value.length > 0;
    }
    if (!valid) {
      this.setState({ validity: ERROR });
    } else {
      this.setState({ validity: VALID });
    }
  };

  textLabel = (name, classes, inputLabel) => {
    return (
      <label htmlFor={name} className={classes}>
        {inputLabel}
      </label>
    );
  };

  render() {
    const {
      additionalClasses,
      cols,
      disabled,
      errorMessage,
      inputLabel,
      name,
      placeholder,
      required,
      rows,
      tag: Tag
    } = this.props;
    const invalid = this.state.validity === ERROR;
    const textAreaProps = { cols, rows };
    const { inputValue } = this.state;

    const labelClassList = classnames("sps-form-group__label", {
      "sps-form-group_label--required": required
    });
    const inputClassList = classnames(
      "sps-form-control",
      // {"is-invalid": this.state.validity === ERROR}
    )
    const useInputGroup = additionalClasses.indexOf("sps-input-group") !== -1;
    const groupClassList = classnames(
      { "sps-form-group": !useInputGroup },
      this.state.interactionState,
      {
        "sps-form-group--error": this.state.validity === ERROR
      },
      {
        "sps-form-group--required": required
      },
      additionalClasses
    );
    return (
      <div className={groupClassList}>
        {inputLabel ? this.textLabel(name, labelClassList, inputLabel) : null}
        <Tag
          ref={this.inputRef}
          defaultValue={inputValue}
          className={inputClassList}
          type="text"
          id={name}
          name={name}
          placeholder={placeholder}
          required={required}
          disabled={disabled}
          onBlur={this.blurHandler}
          onChange={this.changeHandler}
          aria-label={placeholder}
          {...textAreaProps}
        />
        {this.state.validity === ERROR && required ? (
          <small className="error-feedback sps-form-group__feedback--error">{errorMessage}</small>
        ) : null}
      </div>
    );
  }
}

SpsTextInput.propTypes = {
  additionalClasses: PropTypes.array,
  blurCallback: PropTypes.func,
  customValidator: PropTypes.func,
  cols: PropTypes.number,
  disabled: PropTypes.bool,
  equired: PropTypes.bool,
  errorMessage: PropTypes.string,
  inputLabel: PropTypes.string,
  inputValue: PropTypes.string,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  placeholder: PropTypes.string,
  rows: PropTypes.number,
  tag: PropTypes.string
};

SpsTextInput.defaultProps = {
  errorMessage: "This field is required",
  tag: "input",
  additionalClasses: [],
  inputValue: ""
}
