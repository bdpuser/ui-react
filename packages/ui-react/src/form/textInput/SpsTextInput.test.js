import React from "react";
import SpsTextInput from "./SpsTextInput";
import renderer from "react-test-renderer";
import { mount } from "enzyme";

describe("SpsTextInput tests", () => {
  it("renders correctly", () => {
    const output = renderer
      .create(<SpsTextInput inputLabel="first name" name="firstName" />)
      .toJSON();
    expect(output).toMatchSnapshot();
  });
  it("contains input element", () => {
    const wrapper = mount(<SpsTextInput name="firstName" />);
    expect(wrapper.find("input")).toBeDefined();
  });
  it("should have a label", () => {
    const wrapper = mount(
      <SpsTextInput inputLabel="first name" name="firstName" />
    );
    expect(wrapper.find("label").text()).toBe("first name");
  });
  it("should be marked as required when required is a prop", () => {
    const wrapper = mount(
      <SpsTextInput inputLabel="first name" name="firstName" required />
    );
    expect(
      wrapper
        .find("input")
        .html()
        .indexOf("required")
    ).not.toBe(-1);
  });
  it("should be disabled when disabled is a prop", () => {
    const wrapper = mount(
      <SpsTextInput inputLabel="first name" name="firstName" disabled />
    );
    expect(
      wrapper
        .find("input")
        .html()
        .indexOf("disabled")
    ).not.toBe(-1);
  });
  describe("basic validation", () => {
    it("should validate with input", () => {
      const wrapper = mount(
        <SpsTextInput inputLabel="first name" name="firstName" required />
      );
      jest.spyOn(wrapper.instance(), "validateInput");
      wrapper.find("input").simulate("blur", { target: { value: "Bob" } });
      expect(wrapper.instance().validateInput).toHaveBeenCalled();
      expect(wrapper.state("validity")).toBe("valid");
    });
    it("should invalidate with no input", () => {
      const wrapper = mount(
        <SpsTextInput inputLabel="first name" name="firstName" required />
      );
      jest.spyOn(wrapper.instance(), "validateInput");
      wrapper.find("input").simulate("blur", { target: { value: "" } });
      expect(wrapper.instance().validateInput).toHaveBeenCalled();
      expect(wrapper.state("validity")).toBe("error");
      expect(wrapper.find("small.error-feedback").text()).toBe(
        "This field is required"
      );
    });
  });
  describe("callbacks", () => {
    it("should trigger blur callback when provided", () => {
      const callbackSpy = jest.fn();
      const wrapper = mount(
        <SpsTextInput
          inputLabel="first name"
          name="firstName"
          blurCallback={callbackSpy}
          required
        />
      );
      wrapper.find("input").simulate("blur", { target: { value: "Oh boy" } });
      expect(callbackSpy).toHaveBeenCalled();
    });
    it("should trigger custom validator when provided", () => {
      const validatorSpy = jest.fn().mockReturnValue("valid");
      const wrapper = mount(
        <SpsTextInput
          inputLabel="first name"
          name="firstName"
          customValidator={validatorSpy}
          required
        />
      );
      wrapper.find("input").simulate("change", { target: { value: "Bob" } });
      wrapper
        .find("input")
        .simulate("blur", { target: { value: "Chat Noir" } });
      expect(validatorSpy).toHaveBeenCalledWith("Chat Noir");
      expect(wrapper.state("validity")).toBe("valid");
    });
    // TODO: test dirty/pristine
  });
});
