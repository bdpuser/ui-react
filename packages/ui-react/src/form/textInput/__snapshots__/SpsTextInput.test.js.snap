// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpsTextInput tests renders correctly 1`] = `
<div
  className="sps-form-group pristine "
>
  <label
    className="sps-form-group__label"
    htmlFor="firstName"
  >
    first name
  </label>
  <input
    className="sps-form-control"
    defaultValue=""
    id="firstName"
    name="firstName"
    onBlur={[Function]}
    onChange={[Function]}
    type="text"
  />
</div>
`;
